internals/overlord/state/change.go

```diff
@@ -428,7 +429,16 @@ func (c *Change) addNotice() error {
 	opts := &AddNoticeOptions{
 		Data: map[string]string{"kind": c.Kind()},
 	}
-	_, err := c.state.AddNotice(nil, ChangeUpdateNotice, c.id, opts)
+	var extraData map[string]string
+	err := c.Get("notice-data", &extraData)
+	if err == nil {
+		for k, v := range extraData {
+			opts.Data[k] = v
+		}
+	} else if !errors.Is(err, &NoStateError{Key: "notice-data"}) {
+		return fmt.Errorf("could not get notice data from change %s: %v", c.ID(), err)
+	}
```

Comment:
It's kind of odd, but `NoStateError` has an `Is(error) bool` method which checks against `*NoStateError` or `ErrNoState`. So it's simplest (and I think intended) to do `errors.Is(err, ErrNoState)`. [See this test.](https://github.com/canonical/pebble/blob/c330d5f0f37cd7c47b4ec3ce094b39e43fb3f6b4/internals/overlord/state/state_test.go#L1102-L1107)

Also, note we've settled on "cannot ..." for most Pebble errors. Also notice the `%w` verb, which wraps the error, instead of just including it as a string:

```suggestion
	if err == nil {
		for k, v := range extraData {
			opts.Data[k] = v
		}
	} else if !errors.Is(err, ErrNoState) {
		return fmt.Errorf("cannot get notice data from change %s: %w", c.ID(), err)
	}
```

Reply:
Originally I was going to say that in Go we normally do early error returns, which is true, but I wrote it that way and it wasn't actually clearer. So I think the structure you have is fine.